### PR TITLE
Propagate unused columns to distinct on

### PR DESCRIPTION
**Pull request by Tmonster**
Distinct on specifies what columns distinct should be distinct. Currently the optimizer assumes all columns are referenced in distinct on operators. This PR changes the `remove_unused_columns`  optimizer so that references only the needed columns of the distinct on.

Fixes https://github.com/duckdblabs/duckdb-internal/issues/1047

A micro benchmark was added as well
v0.10.0
```
benchmark/micro/distinct/distinct_on_pushdown.benchmark	1	0.219194
benchmark/micro/distinct/distinct_on_pushdown.benchmark	2	0.212928
benchmark/micro/distinct/distinct_on_pushdown.benchmark	3	0.206001
benchmark/micro/distinct/distinct_on_pushdown.benchmark	4	0.206492
benchmark/micro/distinct/distinct_on_pushdown.benchmark	5	0.205673
```

This PR
```
benchmark/micro/distinct/distinct_on_pushdown.benchmark	1	0.047137
benchmark/micro/distinct/distinct_on_pushdown.benchmark	2	0.046034
benchmark/micro/distinct/distinct_on_pushdown.benchmark	3	0.045997
benchmark/micro/distinct/distinct_on_pushdown.benchmark	4	0.045668
benchmark/micro/distinct/distinct_on_pushdown.benchmark	5	0.04746
```

### CODE REVIEW

File: src/optimizer/remove_unused_columns.cpp [hunk 1]
```diff
@@ -165,23 +166,6 @@ void RemoveUnusedColumns::VisitOperator(LogicalOperator &op) {
...
 		return;
-	case LogicalOperatorType::LOGICAL_ORDER_BY:
-		if (!everything_referenced) {
-			auto &order = op.Cast<LogicalOrder>();
-			D_ASSERT(order.projections.empty()); // should not yet be set
-			const auto all_bindings = order.GetColumnBindings();
-
-			for (idx_t col_idx = 0; col_idx < all_bindings.size(); col_idx++) {
-				if (column_references.find(all_bindings[col_idx]) != column_references.end()) {
-					order.projections.push_back(col_idx);
```
Comment (Mytherin) [inline src/optimizer/remove_unused_columns.cpp:176]: I think this is removing an optimization that @lnkuiper added previously - is this intentional?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM - one comment

File: src/optimizer/remove_unused_columns.cpp (hunk 1 above)
Comment (lnkuiper) [reply src/optimizer/remove_unused_columns.cpp:176]: We looked at this together, and managed to re-add the optimization, this time in the appropriate place, the `ColumnLifetimeAnalyzer`.

**Review by lnkuiper — APPROVED** [review]
Looks good to me now, we managed to keep the ORDER BY optimization in

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**